File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <iostream>
#define X first
#define Y second
using namespace std;

vector<int> solution(vector<string> genres, vector<int> plays) {
    unordered_map<string, vector<pair<int,int>>> m;
    unordered_map<string, int> sum;
    vector<int> answer;
    int n = genres.size();
    
    for(int i = 0; i < n; i++){
        m[genres[i]].push_back({plays[i],i});
        sum[genres[i]] += plays[i];
    }
    vector<pair<string, int>> v(sum.begin(), sum.end());
    sort(v.begin(), v.end(), [](pair<string,int> &a, pair<string,int> &b){
                            return a.Y > b.Y;});    

    for(auto x: v){
        sort(m[x.X].begin(), m[x.X].end(), [](pair<int,int> &a, pair<int,int> &b){
            if(a.X == b.X) return a.Y < b.Y;
            return a.X > b.X;});
        for(int i = 0; i < min(2, (int)m[x.X].size()); i++) answer.push_back(m[x.X][i].second);
    }
    
    return answer;
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;
import java.util.stream.Collectors;

class Solution {
    public int[] solution(String[] genres, int[] plays) {
        List<Integer> answer = new ArrayList<>();
        int n = genres.length;
        Map<String, Integer> sum = new HashMap<>();
        Map<String, List<Pair>> all = new HashMap<>();
        
        for(int i = 0; i < n; i++)
            sum.put(genres[i], sum.getOrDefault(genres[i], 0) + plays[i]);
        
        List<Map.Entry<String, Integer>> sumList = sum.entrySet().stream().collect(Collectors.toList());
        
        sumList.sort(Map.Entry.comparingByValue(Comparator.reverseOrder()));
       
        for(int i = 0; i < n; i++) {
            all.put(genres[i], all.getOrDefault(genres[i], new ArrayList<>()));
            all.get(genres[i]).add(new Pair(i, plays[i]));
        }
        
        for(Map.Entry<String, Integer> e : sumList){
            List<Pair> cur = all.get(e.getKey());
            Collections.sort(cur);
            for(int i = 0; i < Math.min(cur.size(), 2); i++)
                answer.add(cur.get(i).idx);
        }
        
        return answer
            .stream()
            .mapToInt(i -> i)
            .toArray();
    }
}

class Pair implements Comparable<Pair> { 
    public int idx;
    public int play;

    Pair(int idx, int play){
        this.idx = idx;
        this.play = play;
    }
    
    public int compareTo(Pair p){
        if(this.play == p.play) return this.idx - p.idx;
        return p.play - this.play;
    }
}